### PR TITLE
[resotocore][feat] Resoto Snapshots: support graph copy

### PR DESCRIPTION
**Pull request by meln1k**
# Description

GraphDB now supports creating a copy of it with a different name

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`
- [x] Document new or updated functionality (someengineering/resoto.com#XXXX)

# Issues Fixed

- Fixes #XXXX

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

File: resotocore/resotocore/db/graphdb.py [hunk 1]
```diff
@@ -1035,7 +1039,68 @@ async def create_update_views(nodes: VertexCollection) -> None:
...
+                        "FOR vertex IN {old_vertex} "
+                        + "LET clean_vertex = UNSET(vertex, '_id', '_rev') "
+                        + "INSERT clean_vertex INTO {new_vertex}"
+                    );
+
+                    db._query(
+                        "FOR edge IN {old_default_edge} "
+                        + "LET clean_edge = UNSET(edge, '_id', '_rev') "
+                        + "LET from = REGEX_REPLACE(edge._from, '^{old_vertex}', '{new_vertex}') "
+                        + "LET to = REGEX_REPLACE(edge._to, '^{old_vertex}', '{new_vertex}') "
```
Comment (aquamatthias) [inline resotocore/resotocore/db/graphdb.py:1080]:
```suggestion
                        + "LET from = CONCAT("{new_vertex}.", PARSE_IDENTIFIER(edge._from)["key"])
                        + "LET to = CONCAT("{new_vertex}.", PARSE_IDENTIFIER(edge._to)["key"])
```
Consider using the AQL builtin function.

File: resotocore/resotocore/db/graphdb.py [hunk 2]
```diff
@@ -1035,7 +1039,68 @@ async def create_update_views(nodes: VertexCollection) -> None:
...
+                        + "INSERT MERGE(clean_edge, {{_from: from, _to: to}}) INTO {new_default_edge}"
+                    );
+
+                    db._query(
+                        "FOR edge IN {old_delete_edge} "
+                        + "LET clean_edge = UNSET(edge, '_id', '_rev') "
+                        + "LET from = REGEX_REPLACE(edge._from, '^{old_vertex}', '{new_vertex}') "
+                        + "LET to = REGEX_REPLACE(edge._to, '^{old_vertex}', '{new_vertex}') "
```
Comment (aquamatthias) [inline resotocore/resotocore/db/graphdb.py:1088]: same here

**Review by aquamatthias — APPROVED** [review]
Nice!

File: resotocore/resotocore/db/graphdb.py [hunk 3]
```diff
@@ -1035,7 +1039,68 @@ async def create_update_views(nodes: VertexCollection) -> None:
...
+    async def copy_graph(self, to_graph: str) -> "ArangoGraphDB":
+        if await self.db.has_collection(to_graph):
+            raise ValueError(f"Graph {to_graph} already exists")
+
+        new_graph_db = ArangoGraphDB(db=self.db, name=to_graph, adjust_node=self.node_adjuster, config=self.config)
+
+        # collection creation can't be a part of a transaction so we do that first
+        # we simply reuse the existing create_update_schema method but do not insert any genesis data
+        async def create_new_collections(new_db: ArangoGraphDB) -> None:
+            await new_db.create_update_schema(init_with_data=False)
```
[the remark this answers is not in the document]
Comment (meln1k) [reply resotocore/resotocore/db/graphdb.py:1054]: Just felt like a better name is nice to have here.

File: resotocore/resotocore/db/graphdb.py [hunk 4]
```diff
@@ -1285,3 +1350,6 @@ async def to_query(self, query_model: QueryModel, with_edges: bool = False) -> T
 
     async def create_update_schema(self) -> None:
         await self.real.create_update_schema()
+
+    async def copy_graph(self, to_graph: str) -> GraphDB:
+        return await self.real.copy_graph(to_graph)
```
Comment (aquamatthias) [inline resotocore/resotocore/db/graphdb.py:1355]: Please emit an event here.

**Pull request merged**